[ML] Add ML memory stats API

droberts195 opened the pull request:
Adds an API that can be used to find out how much memory ML
is permitted to use and is currently using on each node, both
within the JVM heap, and natively, outside of the JVM.

droberts195 committed:
[ML] Add ML memory stats API

Adds an API that can be used to find out how much memory ML
is permitted to use and is currently using on each node, both
within the JVM heap, and natively, outside of the JVM.

elasticsearchmachine commented:
Hi @droberts195, I've created a changelog YAML for you.

droberts195 committed:
Merge branch 'master' into ml_memory_action

droberts195 committed:
Fixing docs

elasticsearchmachine commented:
Hi @droberts195, I've created a changelog YAML for you.

droberts195 committed:
Update docs/changelog/83802.yaml

droberts195 committed:
Fixing docs levels

jgowdyelastic commented:
is the plan for this API to list job ids and their memory usage?

if so, allow the UI to enforce space awareness, this API may need to return stats broken down by job ID, so the kibana server can filter based on jobs the user has permission to see.
It would then be able to reduce the rest into "other jobs" stats without revealing the ids.

If not, i wonder if an overall memory stat would be confusing to the user if all jobs exist in spaces they can't see?

droberts195 commented:
> is the plan for this API to list job ids and their memory usage?

No, it's to provide the numbers to populate this chart:

![image](https://user-images.githubusercontent.com/4185750/140985053-e0686b84-deab-4e50-a577-7a4304aa7020.png)

So it provides numbers per node, not per job.

> If not, i wonder if an overall memory stat would be confusing to the user if all jobs exist in spaces they can't see?

The problem is that whether enough memory exists to start a new job or deploy a trained model depends on jobs in all spaces, not just jobs the user can see.

As I understand it, at the moment that chart is potentially inaccurate because with the numbers calculated in Kibana it's showing total memory for the node but only jobs from spaces the user can see.  So it can give the impression that room is available to start more jobs when really it isn't.  This API solves that problem by showing how much room is left to start new jobs using the exact logic that the backend will use to make this decision.  It's true that there is then a different problem that the sums across all nodes of memory used might be greater than the sums of memory requirements for all jobs the user can see.  Maybe we can get around that by adding a line of explanatory text to say that the chart includes memory used across all spaces?

Users will be able to use this chart to find out that jobs are running in spaces they cannot see, but they won't be able to find out what those jobs are or which spaces they're in.  I think that's OK for different users within the same organisation.  It's probably something that should be reconsidered if we ever wanted to support true multi-tenancy.

cc @darnautov

jgowdyelastic commented:
>As I understand it, at the moment that chart is potentially inaccurate because with the numbers calculated in Kibana it's showing total memory for the node but only jobs from spaces the user can see

It's only showing stats for the current space, it doesn't look across all available spaces.

Ok cool sounds good.

droberts195 committed:
Fixing operator privileges test

droberts195 committed:
Adding some YAML tests

droberts195 committed:
Merge branch 'master' into ml_memory_action

droberts195 committed:
Fix security tests

droberts195 committed:
Adding an integration test on the multi-node cluster

elasticmachine commented:
Pinging @elastic/clients-team (Team:Clients)

sethmlarson reviewed (approved):
Mostly LGTM from a clients perspective, my only recommendation would be to rename from `ml.memory` -> `ml.get_memory_stats` to be more descriptive and match the URL/docs.

lcawl committed:
[DOCS] Sort properties alphabetically

droberts195 committed:
Merge branch 'master' into ml_memory_action

droberts195 committed:
Renaming API spec

benwtrent commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportMlMemoryAction.java:
```diff
@@ -0,0 +1,208 @@
+        ClusterSettings clusterSettings = clusterService.getClusterSettings();
+
+        // Resolve the node specification to some concrete nodes
+        String[] nodeIds = state.nodes().resolveNodes(request.getNodeId());
```
Sneaky little method there. Resolve nodes is a deceptively complicated step!

droberts195 replied:
Yes, it's pretty cool.  I didn't know about the `ml:true` stuff until I started work on this PR.  That's been possible in the node info and node stats APIs since node roles were introduced.

benwtrent replied:
🤯

hendrikmuhs reviewed (approved):
LGTM

The only thing I wonder about: If 1 ML node fails to answer the whole request seems to fail. For a stats endpoint like this, it might be good to be lenient and still return "something". Similar actions have a "node_failures" field. Anyway, this might be something for a follow up PR and should not stop you from merging.

droberts195 commented:
> Similar actions have a "node_failures" field.

This endpoint has a `failures` field that will get populated if there are node failures.  It gets populated by this code, which also generates the `_nodes` summary:

https://github.com/elastic/elasticsearch/blob/12ad399c488f0cc60e19b5e1b29c6d569cb4351a/server/src/main/java/org/elasticsearch/rest/action/RestActions.java#L142-L166

I just realised that I didn't include `failures` in the docs.  But the docs for nodes stats, nodes info, etc., don't include it either.  They just have the same sentence, "If this value is not `0`, a reason for the rejection or failure is included in the response."  Then `failures` is the optional field where those get included.

> If 1 ML node fails to answer the whole request seems to fail.

I think the possible weakness is that the implementation of this action calls _two_ actions that fan out across nodes.  The overall action is a master node action, but with its response formatted as though it were an action that fans out over nodes.  If different nodes fail for the two implementation actions then these aren't necessarily reported perfectly.  The way it works is that the failures from the nodes stats call get propagated as the failures from this endpoint.  Failures from the call that gets the Java model cache size get ignored.  Maybe it would be better to consider a node failed if either call fails for a node.  I will have a closer look at how that could work.

droberts195 committed:
Tightening up timeouts and node failures

1. It's now possible to specify timeout and master_timeout on the
   REST request
2. The timeout is propagated to the requests that fan out to all
   nodes
3. If either internal request fails for a node then the overall
   request is considered to have failed for that node

droberts195 commented:
> Failures from the call that gets the Java model cache size get ignored. Maybe it would be better to consider a node failed if either call fails for a node. I will have a closer look at how that could work.

I made these improvements:

1. It's now possible to specify `timeout` and `master_timeout` on the REST request
2. The `timeout` is propagated to the requests that fan out to all nodes
3. If either internal request fails for a node then the overall request is considered to have failed for that node

droberts195 committed:
Merge branch 'master' into ml_memory_action

droberts195 committed:
Move docs

lcawl committed:
[DOCS] Add new API to ML APIs landing page

lcawl reviewed (approved):
Documentation LGTM
